Add assert statement support

phsilva opened the pull request:
This PR adds assert <expr> statement support to Snek.

When <expr> evaluates to False (per snek_poly_true criteria)
assert will print an AssertionError.

In the future, we could add __debug__ so we can ignore
assertions while __debug__ is false.

**note** this is mostly an experiment to learn more about snek internals... not a very useful language feature but required a bit of internal knowledge.

keith-packard commented on snek-code.c:
```diff
@@ -1177,6 +1179,11 @@ snek_code_run(snek_code_t *code_in)
+			case snek_op_assert:
+				if (!snek_poly_true(snek_a)) {
+					snek_error_0("AssertionError");
+				}
```
Python:
>>> assert True
>>>
Snek:
> assert True
1
>

This can be easily fixed by setting snek_a to SNEK_NULL just after the if() statement here.

keith-packard commented:
Thanks for the fine new feature! This will either need to be conditional or more code squishing will be needed to fit it into the smallest platforms (currently amega 32u4). See also a minor comment I made inline to keep assert from returning a value when used at the REPL.

keith-packard commented:
Otherwise, this patch looks fantastic. Thanks especially for adding tests to make sure it keeps working in the future.

keith-packard commented:
oh, if you get excited, adding documentation to the manual about this would be a nice addition.

phsilva commented:
Hi @keith-packard, added documentation a fix the REPL issue.

The only missing part is conditional compiling. WIll get it soon.

phsilva commented:
The conditional compilation is a little trickier than I expected as I have to add it conditionally even in the grammar. Will do some tests to see if leaving the grammar but removing my vm code will make it smaller enough.

I am afraid tough that the atmega32y4 (tested with itsybitsy5v) is already too big in master (maybe my avr-gcc version or something else). Git this when compiling from master.

```
/usr/local/opt/avr-binutils/bin/avr-ld: region `text' overflowed by 2542 bytes
collect2: error: ld returned 1 exit status
```

keith-packard commented:
[> The conditional compilation is a little trickier than I expected as I have to add it conditionally even in the grammar. Will do some tests to see if leaving the grammar but removing my vm code will make it smaller enough.

Just for you, I've added a preprocessor to lola so we can have conditional code in the grammar.
Usage is a bit funky, but I think adequate for this purpose.

You can make portions of the grammar conditional based on defines provided on the command line.

```
foo	: bar
    	{ppsym
    		| bletch
    	}
    		;
```

If you add  '-Dppsym' on the command line, then the '| bletch' production will be included, otherwise
it will be skipped.

This is in version 1.2 of lola, which I just pushed.

> I am afraid tough that the atmega32y4 (tested with itsybitsy5v) is already too big in master (maybe my avr-gcc version or something else). Got this when compiling from master.

Hrm. I had neglected to push a lola change that reduced the size of another enum, but I don't think that would account for such a large change. I'm using avr-gcc version 5.4.0.

> /usr/local/opt/avr-binutils/bin/avr-ld: region `text' overflowed by 2542 bytes
> collect2: error: ld returned 1 exit status

That's a big difference. Even including the 'assert' code, I'm only 60 bytes over on 32u4 and 26 bytes over on duemilanove. Go ahead and try the most recent lola bits and see what you get.

phsilva commented:
The binary size was really an issue with my avr-gcc version (was using 9.0 or something), moved it back to avr-gcc 5 and results were consistent.

Added conditional compilation support around assert. Had to add support for in snek-builtin.py too (see b819195970496a8854e9d5c83467329583b9ccb9).

It looks good but using the new condition the assert stopped. I couldn't figure out yet why. All the code seems to be defined correctly when I defined SNEK_ASSERT but I get a SyntaxError when try to use it, maybe something yet missing in lola (which I updated), don't know, will check.

As usual, an apparently innocuous new statement is giving nice opportunities to learn about snek.

keith-packard commented on snek-keyword.builtin:
```diff
@@ -14,7 +14,11 @@
 global, GLOBAL
-assert, ASSERT
+
+#ifdef SNEK_ASSSERT
```
One too many 'S' here means assert is never a keyword :-)

keith-packard commented on snek.defs:
```diff
@@ -76,7 +76,16 @@ SNEK_WARNINGS = \
 	-Wimplicit-fallthrough
 
-SNEK_BASE_CFLAGS = -D_DEFAULT_SOURCE -I$(SNEK_ROOT) -DSNEK_VERSION='"$(SNEK_VERSION)"'
+SNEK_ASSERT ?=
```
I think you want the default value for SNEK_ASSERT to be 1 instead of nothing?

keith-packard commented:
Glad you  resolved the size issue. Sad to hear the latest avr-gcc makes much larger code than the old version.

With the two minor issues resolved, I've got assert working here. Let me know when you've updated your branch and I'll merge it in. Thanks again for this fine new feature!
